### PR TITLE
feat: Create function which translates profiles defined for one zone set to another zone set

### PR DESCRIPTION
**Pull request by merrielle**
[Pull Request doc](https://breakthrough-energy.github.io/docs/user/git_guide.html#d-pull-request)

### Purpose
Addresses https://github.com/Breakthrough-Energy/PreREISE/issues/259
"Create function which translates profiles defined for one zone set to another zone set"

**DEMO JUPYTER NOTEBOOK W/ DATA:**
https://www.dropbox.com/s/srbbkxirykug843/zone_translation.ipynb?dl=0

### What the code is doing
`translate_zone_set()` has three key steps:
- Overlay map of `prev_zones` with `new_zones`
- Pivot table to make it a matrix of `prev_zones` x `new_zones`
- Scale the matrix such that each row sums to 1, then return it

Also included some plotting and results analysis functions. Not sure if these belong in PostREISE?

- `plot_zone_map()` can do both interactive maps and static PNGs
- `filter_interesting_zones()` takes the output of `translate_zone_set()` and filters it to only show when a single zone has been distributed to several new zones

### Testing
TODO

### Where to look
`translate_zone_set()` is the most important

### Usage Example/Visuals
Example of `plot_zone_map()` using Liming's BA shapefiles:
liming_ba_map

Example of `plot_zone_map()` using Merrielle's BA shapefiles:
merrielle_ba_map

Example of `filter_interesting_zones()`, showing translation from Liming's BA shapefiles to Merrielle's BA shapefiles:
**NOTE**: We're rounding the results to three decimal places to make them more readable, so not all rows add up to 1.

Screen Shot 2022-03-01 at 7 25 10 PM

### Time estimate
30 mins?

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -16,3 +16,6 @@ black
 pytest
 coverage
 pytest-cov
+folium
+geopandas
+rtree
```
Comment (danielolsen) [inline requirements.txt:22]: We'll want to add these to the Pipfile so that we can sync all necessary dependencies.

File: requirements.txt [hunk 2]
```diff
@@ -16,3 +16,6 @@ black
 pytest
 coverage
 pytest-cov
+folium
```
Comment (danielolsen) [inline requirements.txt]: It would be nice if we could add this as an optional requirement (and avoiding importing if `interactive=False`), since it seems like it's not strictly necessary.

File: prereise/gather/translate_zones.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        prev_to_new_matrix["sum"].round(3) > 1
+    ]
+    if len(sum_greater_than_one > 0):
+        print(
+            f"WARNING: Detected rows with a sum greater than one. This most likely means that several {name_new} zones overlap each other."
+        )
+        if verbose:
+            print("\nROWS WITH SUM GREATER THAN ONE:")
+        if verbose:
+            print(sum_greater_than_one.round(3))
```
Comment (danielolsen) [inline prereise/gather/translate_zones.py]: Can these be combined into one `if` block?

File: prereise/gather/translate_zones.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+    )
+    prev_to_new_matrix["to_distribute"] = 1 - prev_to_new_matrix["sum_without_none"]
+    prev_to_new_matrix["scale_factor"] = (
+        1 + prev_to_new_matrix["to_distribute"] / prev_to_new_matrix["sum_without_none"]
+    )
+
+    if verbose:
+        print("\nMATRIX BEFORE SCALING")
+    if verbose:
+        print(prev_to_new_matrix.round(3))
```
Comment (danielolsen) [inline prereise/gather/translate_zones.py]: Same comment here.

File: prereise/gather/translate_zones.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        columns = scaled_matrix.columns
+        for zone in list(new_zones[name_new]):
+            if not zone in columns:
+                scaled_matrix[zone] = 0
+
+    if verbose:
+        print(
+            "\nINTERESTING ZONES: rows and cols that include values that are not 1 or 0"
+        )
+        with pd.option_context("display.max_rows", None, "display.max_columns", None):
```
Comment (danielolsen) [inline prereise/gather/translate_zones.py:218]: Nice

File: prereise/gather/translate_zones.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
+import folium
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+from folium.features import DivIcon
+
+# IMPORTANT: rtree is required for overlay function, but not imported
```
Comment (danielolsen) [inline prereise/gather/translate_zones.py:5]: I'm guessing we'll get some sort of Exception if we try to use `overlay` without `rtree`? Can we add a `try`/`except` block around the `overlay` call that provides a more direct warning to the user if `rtree` is missing?

**Comment by BainanXia** [comment]
> Also included some plotting and results analysis functions. Not sure if these belong in PostREISE?

I think put these plot functions closer to its main context is good. The reason we have most visualizations in PostREISE is because the package is designed for post analysis of simulation results rather than a place where we should put all visualizations in. Open to discussion.

Not sure whether we will use functionalities here in scenario analysis in the future though. But if that's the case, we will definitely need more than geographical definitions, thus more complicated plots in PostREISE will be implemented anyways.

File: prereise/gather/translate_zones.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+
+
+def plot_zone_map(gdf, interactive=False):
```
Comment (BainanXia) [inline prereise/gather/translate_zones.py]: The map looks great in the demo! Not sure whether we would like the plot parameters in this function to be hardcoded (size, color, etc.) or not, given we might want to look into a smaller area rather than the entire nation some day. Maybe this is future work.

**Comment by BainanXia** [comment]
A higher level question for the team: should we put support modules like this in `gather` or should we create another folder, say `utility` next to `gather` as we did in `PowerSimData`?

File: prereise/gather/translate_zones.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+    :return: (*pandas.DataFrame*) -- subset of original matrix
+    """
+    df_copy = df.round(rounding)
+    df_copy = df_copy.replace({1.0: 0.0})
+    df_copy.loc["sum"] = df_copy.sum()  # sum each col
+    df_copy["sum"] = df_copy.sum(axis=1)  # sum each row
+
+    interesting_zones = df_copy.loc[df_copy["sum"] > 0, df_copy.loc["sum"] > 0]
+    interesting_zones = interesting_zones.drop(index=["sum"], columns=["sum"])
+    interesting_zones = interesting_zones.replace({0.0: "-"})
```
Comment (BainanXia) [inline prereise/gather/translate_zones.py]: Maybe we can do following here:
```python
interesting_zones = df_copy.loc[(df_copy != 0).any(axis=1), (df_copy != 0).any()].replace({0.0: "-"})
```

File: prereise/gather/translate_zones.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+
+    return interesting_zones
+
+
+def format_zone_df(df, name):
+    """copies df, checks if it contains valid geometry, and adds col for zone area
```
Comment (BainanXia) [inline prereise/gather/translate_zones.py]: `valid` -> `invalid`?

File: prereise/gather/translate_zones.py [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+    is_valid_df = df_copy.geometry.is_valid.to_frame()
+    invalid_geom = list(is_valid_df.loc[is_valid_df[0] == False].index)
+    if len(invalid_geom) > 0:
+        print(f"WARNING: {name} contains invalid geometries: {invalid_geom}")
+
+    df_copy.index.name = name
+    df_copy = df_copy.reset_index()
+
+    # get zone area
+    df_copy[f"{name}_area"] = df_copy["geometry"].area
```
Comment (BainanXia) [inline prereise/gather/translate_zones.py:94]: What would be the `area` value if the `geometry` is invalid?

File: prereise/gather/translate_zones.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+    :param geopandas.geodataframe.GeoDataFrame df: GeoDataFrame where
+        index = zone names, columns = ['geometry']
+    :param str name: the name of the data set
+    :return: (*geopandas.geodataframe.GeoDataFrame*) -- the formatted df
+    """
+    df_copy = df.copy()
+
+    # Check for invalid geometries
+    is_valid_df = df_copy.geometry.is_valid.to_frame()
+    invalid_geom = list(is_valid_df.loc[is_valid_df[0] == False].index)
```
Comment (BainanXia) [inline prereise/gather/translate_zones.py]: Not sure whether we can do
```python
invalid_geom = df_copy.loc[~df_copy.geometry.is_valid].index.tolist()
```
here.

File: prereise/gather/translate_zones.py [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+
+
+def translate_zone_set(
+    prev_zones, new_zones, name_prev="prev_zones", name_new="new_zones", verbose=False
+):
+    """returns matrix of prev_zones to new_zones. Scales rows proportionally so
+    that each row adds up to 1
+    NOTE: We recommend using EPSG:4269 as your CRS (coordinate reference system).
+    It triggers a warning but it is able to handle invalid geometries, unlike
+    Mercator Projections (EPSG:3395).
```
Comment (danielolsen) [inline prereise/gather/translate_zones.py]: Do we really want to be recommending `EPSG:4269` here? I think this one is defined in lat/lon, but if we're doing area calculations then I think we want an equal-area projection. As a test, we can look at a lat/lon rectangle, with its bottom edge at the equator, and translate it into two sub-areas, one for the 'lower' half and one for the 'upper' half. We should see that the 'lower' half is larger.

As a side note: do we want to handle invalid geometries?

Comment (BainanXia) [reply prereise/gather/translate_zones.py]: A follow up question regarding invalid geometries: how do we define invalid geometries in our situation? It seems any shape doesn't form a polygon is treated as invalid from the [documentation](https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoSeries.is_valid.html), however, we can have bowtie like shapes of areas (BAs or utilities?), right?

File: prereise/gather/translate_zones.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+    # if old zones aren't completely covered by new zones, we have < 100% coverage
+    # Note: we could technically simplify this to just scale_factor = 1/(sum - none) but
+    # the extra cols are nice for verbose mode
+    prev_to_new_matrix["sum_without_none"] = (
+        prev_to_new_matrix["sum"] - prev_to_new_matrix["none"]
+    )
+    prev_to_new_matrix["to_distribute"] = 1 - prev_to_new_matrix["sum_without_none"]
+    prev_to_new_matrix["scale_factor"] = (
+        1 + prev_to_new_matrix["to_distribute"] / prev_to_new_matrix["sum_without_none"]
+    )
```
Comment (BainanXia) [inline prereise/gather/translate_zones.py:192]: Thanks for putting down detailed explanation in doc strings! Just copying our discussions from Slack to here regarding how to redistribute `over counts` to make rows summing up to 1:

> Another approach to bring it down to 1 is looking at the overlaps among columns and distribute the overlap portion uniformly cross columns that causing duplications rather than all areas.

This will make things messy for sure but gives a relative better projection when dealing with self overlaps. I'm good with the current implementation. Just posting the idea here for future references.

File: prereise/gather/translate_zones.py [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+        columns=["none", "sum", "sum_without_none", "to_distribute", "scale_factor"]
+    )
+
+    # Add any rows that got dropped because they do not overlap. Fill with zeros.
+    # NOTE: not sure if this will get slow with larger datasets
+    if len(scaled_matrix) < len(prev_zones):
+        index = scaled_matrix.index.values
+        for zone in list(prev_zones[name_prev]):
+            if not zone in index:
+                scaled_matrix.loc[zone] = 0
```
Comment (BainanXia) [inline prereise/gather/translate_zones.py]: Maybe we can do following here:
```python
scaled_matrix.loc[list(set(prev_zones[name_prev]) - set(scaled_matrix.index))] = 0
```
Similar for adding columns below.

File: prereise/gather/translate_zones.py (hunk 6 above)
Comment (merrielle) [reply prereise/gather/translate_zones.py:5]: Sure! That sounds like a good idea.

File: requirements.txt (hunk 2 above)
Comment (merrielle) [reply requirements.txt]: Honestly it might be better for me to figure out how to do this in Bokeh so we don't need the extra library at all

File: prereise/gather/translate_zones.py (hunk 6 above)
Comment (merrielle) [reply prereise/gather/translate_zones.py:5]: Actually, the exception from geopandas is very explicit and I don't want to accidentally mask other issues that may come up. Is it okay if I leave this as-is?

Comment (danielolsen) [reply prereise/gather/translate_zones.py:5]: Nice! I'm too used to unhelpful exceptions. If the `geopandas` one is already good, we can keep it.

File: prereise/gather/translate_zones.py (hunk 12 above)
Comment (merrielle) [reply prereise/gather/translate_zones.py]: In the case of a bowtie, the shape is better represented with a MultiPolygon. MultiPolygons are also used for BAs with several discrete areas.

File: prereise/gather/translate_zones.py (hunk 13 above)
Comment (merrielle) [reply prereise/gather/translate_zones.py:192]: I think you're right about this, but it will be quite a bit more work if we want to check for three or more BAs overlapping. To handle three or more simultaneous overlapping BAs we need to recursively check for overlaps until there are none left.

We also need to think about how we handle gaps. For now I think distributing demand proportionally works best to fill in any gaps in the new shapefiles.

Here's an article I've been reading about gaps and overlaps:
https://medium.com/@achm.firmansyah/an-approach-for-checking-overlaps-and-gaps-in-polygons-using-geopandas-ebd6606e7f70

File: prereise/gather/translate_zones.py (hunk 12 above)
Comment (BainanXia) [reply prereise/gather/translate_zones.py]: So `is_valid()` will return `True` in this case?

File: prereise/gather/translate_zones.py (hunk 14 above)
Comment (merrielle) [reply prereise/gather/translate_zones.py]: I believe assigning with loc only works if you do one row at a time. If one is missing we get:
`"None of [Index(['c'], dtype='object', name='gray_zones')] are in the [index]"`

Using a for loop for now, but let me know if you know a better way!
```python
new_rows = list(set(prev_zones[name_prev]) - set(scaled_matrix.index))
for row in new_rows:
    scaled_matrix.loc[row] = 0
```

Comment (danielolsen) [reply prereise/gather/translate_zones.py]: `pandas.concat` will let you add an arbitrary number of rows at once, but doesn't modify inplace like a `loc` or `append` addition.

Comment (BainanXia) [reply prereise/gather/translate_zones.py]: Yeah, you are right. Adding multiple rows in-place in one shot is not supported. Other than using a loop, using `append`/`concat` should work as well, not in-place operation though:
```python
scaled_matrix  = pd.concat([scaled_matrix, pd.DataFrame(0, index=new_rows, columns=scaled_matrix.columns)])
```

**Comment by merrielle** [comment]
Looks like using Pseudo-Mercator does actually affect the results -- mostly by a couple thousandths. Included a quick screenshot of partial new results on top and partial old results on the bottom.

Screen Shot 2022-03-16 at 4 53 00 PM

File: prereise/gather/translate_zones.py (hunk 12 above)
Comment (merrielle) [reply prereise/gather/translate_zones.py]: If it's a bowtie as a multipolygon then it's valid. If it's a single polygon that crosses itself, it's invalid.

Comment (danielolsen) [reply prereise/gather/translate_zones.py]: Do we see any bowties in the current draft of the BA shapes?

Comment (BainanXia) [reply prereise/gather/translate_zones.py]: I see. So it depends on the object type within the inputs. Not sure whether our existing shapefile runs into this issue though.

File: requirements.txt (hunk 2 above)
Comment (merrielle) [reply requirements.txt]: Update: I'd like to just ditch the interactive graph. It'd take me a while to refresh my bokeh knowledge and I feel like since we already have a static graph the interactive one is not a huge value add anyway.

File: prereise/gather/translate_zones.py (hunk 12 above)
Comment (merrielle) [reply prereise/gather/translate_zones.py]: Okay, so I've tested bowties specifically, and it looks like the area is calculated incorrectly. I've searched for a bunch of different ways to fix invalid polygons, and it seems like there is no single reliable way. I think it's best to throw an exception if any invalid polygon is detected.

Here's a bowtie with "zero" area. :(
Screen Shot 2022-03-17 at 2 35 58 PM

Comment (merrielle) [reply prereise/gather/translate_zones.py]: Here's the same bowtie as a multipolygon with the area calculated correctly:
Screen Shot 2022-03-17 at 2 49 35 PM

Note that I only found one invalid geometry in Liming's shapefiles -- most of the BAs were already multipolygons.

Comment (danielolsen) [reply prereise/gather/translate_zones.py]: Is there any sort of helpful utility that converts bowties to multipolygons?

Comment (merrielle) [reply prereise/gather/translate_zones.py]: I tried a couple different options from stack overflow, and the best I've seen is `make_valid()`. I had looked it up previously, and I thought I saw someone mention some issues with it, but I can't find that post anymore. I tested it with two types of invalid polygons (a bowtie and a flag) and it seemed to work. If we want to use it, we need to add shapely 1.8 as a requirement.

Comment (merrielle) [reply prereise/gather/translate_zones.py]: Here's a screenshot of my tests if you're interested. I'm not sure if screenshots like this are helpful but I figure it's less work for you than sending over a notebook.
Screen Shot 2022-03-20 at 3 16 44 PM

Comment (BainanXia) [reply prereise/gather/translate_zones.py]: > I tried a couple different options from stack overflow, and the best I've seen is `make_valid()`. I had looked it up previously, and I thought I saw someone mention some issues with it, but I can't find that post anymore. I tested it with two types of invalid polygons (a bowtie and a flag) and it seemed to work. If we want to use it, we need to add shapely 1.8 as a requirement.

Not sure whether the issue you saw is among the replies of [this](https://stackoverflow.com/questions/20833344/fix-invalid-polygon-in-shapely) post. One of them mentioned dependency conflicts due to shapely 1.8a3 at that time wasn't a stable release, which shouldn't be a problem given 1.8.1 is released now. While, the very last post mentioned a problem that `make_valid` might return a multi-polygon of length greater than 1, which seems to be the normal behavior as expected (not sure whether my read is correct though)? All in all, I think this solution should work unless running into some corner cases which we might be able to fix it case by case.

Comment (merrielle) [reply prereise/gather/translate_zones.py]: I wonder if we should be trying to fix other people's input data though. Maybe I can throw an error if there's invalid geometry but include a separate helper function that people can use to fix their shapefiles? To me it seems like cleaning the data is a separate concern.

Comment (danielolsen) [reply prereise/gather/translate_zones.py]: I think an Exception that tells the user why their geometries can't be ingested as passed is fine within this PR, assuming we've got a set of valid geometries that comes out of #270 so that our use case is fully automated.

**Comment by merrielle** [comment]
@BainanXia @danielolsen
Aside from pytest failing (it passes locally so I'm investigating) I think this is ready for a final pass. I've addressed all the PR comments, added tests, and added a helper function `make_zones_valid()` that can be used to clean any imported gdf.

Regarding Bainan's comment about changing how we distribute overlaps:
> Another approach to bring it down to 1 is looking at the overlaps among columns and distribute the overlap portion uniformly cross columns that causing duplications rather than all areas.

I've been investigating this more closely and there are many many edge cases that we may or may not want to handle. The rabbit hole goes deep. I'll send you a summary of some of my findings and address this in a future PR.

**Comment by danielolsen** [comment]
Are the tests failing because we add geopandas to the `Pipfile` but we haven't generated a new `Pipfile.lock`?

**Comment by merrielle** [comment]
Possibly! Updating it now.

**Review by danielolsen — APPROVED** [review]
Thanks!
